File: 0213.house_robber_II/solution.go
package main

func rob(nums []int) int {
	length := len(nums)
	if length == 0{
		return 0
	}
	if length < 3{
		return max(nums)
	}
	return max([]int{sub(nums[1:]), sub(nums[:length-1])})
}


func sub(nums []int) int{
	dp := make([]int, len(nums))
	dp[0] = nums[0]
	dp[1] = max([]int{nums[0], nums[1]})
	for i := 2;i<len(nums);i++{
		dp[i] = max([]int{dp[i-1], dp[i-2] + nums[i]})
	}
	return max(dp)
}

func max(nums []int) int {
	v := nums[0]
	for _, n := range nums{
		if n > v{
			v = n
		}
	}
	return v
}

File: 0213.house_robber_II/solution.py
from typing import List


class Solution:
    def rob(self, nums: List[int]) -> int:
        length = len(nums)
        if length == 0:
            return 0
        if length < 3:
            return max(nums)

        def _rob(m):
            dp = [0] * len(m)
            dp[0] = m[0]
            dp[1] = max(m[0], m[1])
            for i in range(2, len(m)):
                dp[i] = max(dp[i-1], dp[i-2] + m[i])
            return max(dp)

        # 因为是环形排列，所以取移除第一个和最后一个的最大值即可
        return max(_rob(nums[1:]), _rob(nums[:-1]))
